crates/red_knot_python_semantic/mdtest.py

```diff
@@ -0,0 +1,200 @@
+"""A runner for Markdown-based tests for Red Knot"""
+# /// script
+# requires-python = ">=3.11"
+# dependencies = [
+#     "rich",
+#     "watchfiles",
+# ]
+# ///
+
+import json
+import os
+import subprocess
+from pathlib import Path
+from typing import Literal, Never
+
+from rich.console import Console
+from watchfiles import Change, watch
+
+CRATE_NAME = "red_knot_python_semantic"
+CRATE_ROOT = Path(__file__).resolve().parent
+MDTEST_DIR = CRATE_ROOT / "resources" / "mdtest"
+
+
+class MDTestRunner:
+    mdtest_executable: Path
+    console: Console
+
+    def __init__(self) -> None:
+        self.console = Console()
+
+    def _run_cargo_test(self, message_format: Literal["human", "json"]) -> str:
+        return subprocess.check_output(
+            [
+                "cargo",
+                "test",
+                "--package",
+                CRATE_NAME,
+                "--no-run",
+                "--color=always",
+                "--message-format",
+                message_format,
+            ],
+            cwd=CRATE_ROOT,
+            env=dict(os.environ, CLI_COLOR="1"),
+            stderr=subprocess.STDOUT,
+            text=True,
+        )
+
+    def _recompile_tests(
+        self, status_message: str, message_on_success: bool = True
```

Comment:
nit: I generally prefer to make most `bool` flags keyword-only; it makes callsites more readable

```suggestion
        self, status_message: str, *, message_on_success: bool = True
```